src/mastra/workflows/logCoreAnalyzer/utils.ts

```
@@ -1,89 +1,192 @@
-  return Math.ceil(text.length * tokensPerChar);
+  // Sample three windows to reduce bias (head/mid/tail).
+  const half = Math.floor(SAMPLING_WINDOW_SIZE / 2);
+
+  const startStart = 0;
+  const midCenter = Math.floor(len / 2);
+  const endEnd = len;
+
+  const head = text.slice(startStart, Math.min(SAMPLING_WINDOW_SIZE, len));
+  const mid = text.slice(
+    Math.max(0, midCenter - half),
+    Math.min(len, midCenter + half)
+  );
+  const tail = text.slice(Math.max(0, endEnd - SAMPLING_WINDOW_SIZE), endEnd);
+
+  // Tokenize each window (small, fixed cost).
+  const headTokens = countTokens(head);
+  const midTokens = countTokens(mid);
+  const tailTokens = countTokens(tail);
+
+  const sampledChars = head.length + mid.length + tail.length;
+  const sampledTokens = headTokens + midTokens + tailTokens;
+
+  // Guard against division by zero (shouldn't happen, but be safe).
+  if (sampledChars === 0) return 0;
+
+  const tokensPerChar = sampledTokens / sampledChars;
+  return Math.ceil(len * tokensPerChar);
 };
 
+/**
+ * Exact token count for a given text (uses your tokenizer).
+ * @param text - Text to count tokens for
+ * @returns Number of tokens
+ */
+export const countTokens = (text: string): number => encode(text).length;
+
 /**
  * Validate token count against configured limits
  * @param text - Text to validate token count for
  * @returns Estimated number of tokens
  * @throws Error if token count exceeds limit
  */
-export function validateTokenLimit(text: string): number {
+/**
+ * Create a token limit error message
+ * @param estimatedTokens Actual number of tokens
+ * @param maxTokens Maximum allowed tokens
+ * @returns Formatted error message
+ */
+const createTokenLimitError = (
+  estimatedTokens: number,
+  maxTokens: number
+): Error =>
+  new Error(
+    `Token limit constraint violated: Estimated ~${estimatedTokens} tokens, which exceeds the configured maximum of ${maxTokens} tokens`
+  );
```

[nit] I think you should just throw the error instead of adding a util function. Also looks like this was inserted in between jsdoc comments